### PR TITLE
Faraday Website

### PR DESCRIPTION
**Pull request by iMacTia**
## Description
Introduces a GitHub-pages-compatible website based on Jekyll and the Type theme.
The website will also be used to reorganise and extend the README content with tutorials, mini-guides and general documentation.
This will NOT replace the YARD-generated API Documentation.
Supersedes #970.

## Todos
List any remaining work that needs to be done, i.e:
- [x] Complete moving README content into separate pages
- [x] Add explanation in README on how to contribute to the website (Jekyll instructions)
- [x] Add more documentation pages
  - [x] Team
  - [x] Introduction
  - [x] Finish Middleware pages
    - [x] Multipart
    - [x] UrlEncoded
    - [x] Retry
    - [x] Instrumentation
    - [ ] List and links to popular 3rd party middleware
  - [ ] SSL
  - [ ] Parallel Requests
- [x] Add some navigation between documentation pages (next/prev/menu)
- [x] Add some branding/styles to website
- [ ] First Blog page (article on v1.0 or new website?)
- [x] Correct links to Faraday website (currently https://???)

### CODE REVIEW

**Comment by olleolleolle** [comment]
Fantastic work!

**Comment by olleolleolle** [comment]
@iMacTia Could you build out the TODO list in the Description field with some more pages/articles to include?

**Comment by olleolleolle** [comment]
I updated the wiki - avoid formatting errors, dead links.

Perhaps for the website we want 

- SSL tips page
- 3rd-party middleware links
- How to use parallel requests in the Typheous adapter page?

**Comment by iMacTia** [comment]
@olleolleolle that makes sense, I’ll drop a detailed TODO list when I have a chance, for now I’m making a page for each out-of-the-box middleware like in my last commit, if you want to help 👍

**Comment by olleolleolle** [comment]
@iMacTia One thing I began doing was to make a `git shortlog -e -s`-based Team page - I used a `.mailmap` file to combine a few known duplicates, and then have a full roster of people involved.

If we want a list of all contributors, here is an automated way to do that.

<details>

<summary>How I used a mailmap file to combine committer identities</summary>

# Create a summary file of all contributors

With this `.mailmap` file in the repo root: (I dotted out the email address domains)

```
Rick Olson <technoweenie@...>
Dor Shahaf <dor@...> Dor Shahaf <SoapSeller@...>
Mislav Marohnić <mislav.marohnic@...>
Mattia Giuffrida <giuffrida.mattia@...> Mattia <iMacTia@...>
Mattia Giuffrida <giuffrida.mattia@...> iMacTia <giuffrida.mattia@...>
```

You can ask git to generate a list of commit summaries for contributors:

```
git shortlog -esn
```

This lists everyone with number of contributions in the first column:

```
   420  Rick Olson <technoweenie@...>
   276  Mislav Marohnić <mislav.marohnic@...>
   161  Erik Michaels-Ober <sferik@...>
   144  Mattia Giuffrida <giuffrida.mattia@...>
   112  Olle Jonsson <olle.jonsson@...>
    29  Zack Hobson <zack@...>
```

This list can be cut down to size: lose the first column, then split on delimiter `<`, and pick "field 0".

```
git shortlog -esn | cut -c8- | cut -d '<' -f1
```




</details>

**Comment by iMacTia** [comment]
@olleolleolle that's amazing! Indeed the Team Page is something I was planning and you might have noticed in the main menu 😃! The idea would be to pay tribute to all the maintainers and members of the core team that helped building Faraday over the time, plus a quick note on the current active maintainers (so people know who to get in touch with)

File: docs/introduction/basics.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+## POST ##
+
+# Parameters for POST requests are automatically put in the body as
+# www-form-urlencoded.
+conn.post '', { name: 'Maguro' }
+# => POST "name=maguro" to http://sushi.com/nigiri
+
+# To post as JSON instead of www-form-urlencoded, set the request header
+conn.post do |req|
```
Comment (technoweenie) [inline docs/introduction/basics.md]: Faraday doesn't automatically convert hashes to json without the `json` middleware in `faraday_middleware`. Doing this correctly looks like this:

```ruby
require 'faraday'
require 'faraday_middleware'
f = Faraday.new do |b|
  b.request :url_encoded
  b.request :json
  b.adapter Faraday.default_adapter
end

f.post $url, {data: "json"}, "Content-Type" => "application/json"
```

If the `json` middleware is not loaded, this is what you get:

```
>> Faraday.post $url, {ctype: "json"}, 'Content-Type' => 'application/json'
NoMethodError: undefined method `bytesize' for {:ctype=>"json"}:Hash
```

What do you think about moving this JSON bit to a page specifically for the `json` middleware? This section could then be clear that Faraday only handles url/form/multipart encoding for you.

I also tend to prefer handling JSON encoding explicitly:

```ruby
f.post $url, {data: "json"}.to_json, "Content-Type" => "application/json"
```

File: docs/introduction/streaming.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    streamed << chunk
+  end
+end
+
+# Joins all response chunks together 
+streamed.join
+```
+
+The `on_data` streaming is currently only supported by the `Net::HTTP` adapter. Other adapters
+will be support this in the future.
```
Comment (technoweenie) [inline docs/introduction/streaming.md]: Supporting this in other adapters sounds like a promise we won't keep :)

**Comment by iMacTia** [comment]
@technoweenie I also looked mostly at the bundler site and found it a great leading example 😃!
Completely agree with your observations above, I have mostly moved content from the README and focused on reorganising it into separate pages, trying to apply some style. I'm sure I missed plenty of stuff but I'll review keeping your points in mind, feel free to point out anything that seems off.
I'll fix your comments to begin with, then run a full check of each page 👍

File: docs/introduction/basics.md (hunk 1 above)
Comment (iMacTia) [reply docs/introduction/basics.md]: Hey, yes I know the JSON middleware is not included in Faraday, but if you check my example it sets the body as a stringified JSON, so it should work out of the box.
This is of course reductive and cumbersome, but that's we have the JSON middleware in `faraday_middleware`, although I agree that deserves its own page (and arguably on another website?)

Comment (technoweenie) [reply docs/introduction/basics.md]: Ahh, must've missed those quote, you got me :) I wonder if it should show generic examples of setting the req body to a string or io reader instead, even if the sample code is still sending raw json. What do you think?

1. Send form/multipart as ruby hash
2. Send string resp body with content type
3. Send io reader resp body with content type

Comment (iMacTia) [reply docs/introduction/basics.md]: Totally makes sense 👍! Thanks

**Comment by iMacTia** [comment]
@technoweenie title is now fine, although we're using my fork of the theme.
We can switch back to the upstream theme once my PR is merged 👍 
That makes the order of the nav menu the only outstanding issue

**Comment by iMacTia** [comment]
PR has now been merged https://github.com/rohanchandra/type-theme/pull/88 🎉

**Comment by technoweenie** [comment]
https://github.com/rohanchandra/type-theme/pull/90 added the ability to sort the site nav items.

Homepage___Faraday

**Pull request merged**